### PR TITLE
Mailroom part 2 (trigrams still WIP)

### PR DESCRIPTION
**Pull request by theseanhodges**

### CODE REVIEW

File: students/shodges/lesson04/mailroom.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+def send_thank_you():
+    while True:
+        donor = input('Please enter a donor name: ')
+        if donor == 'quit':
+            break
+        elif donor == 'list':
+            for item in donors.keys():
+                print(item)
+        else:
+            if not donor in donors.keys(): donors[donor] = []
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: this should be two separate lines

File: students/shodges/lesson04/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            amount = input('Please enter a donation amount: ')
+            try:
+                donors[donor].append(float(amount))
+            except ValueError:
+                break
+            total = 0
+            for gift in donors[donor]: total += gift
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: should be two separate lines; however use `sum(donors[donor])` here

File: students/shodges/lesson04/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            break
+
+def generate_report():
+    print('{:24} | {:10} | {:10} | {:12}'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print('-'*68)
+    donor_report = [] # not a ton of value in converting this to a list.  it made things a bit more complicated...
+    for donor in donors.keys():
+        total = 0
+        for amount in donors[donor]:
+            total += amount
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: same here, use `sum()`

File: students/shodges/lesson04/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    letter_dir = Path(input('Please specify a directory to save letters in: '))
+    if not letter_dir.is_dir():
+        print('{} is an invalid directory; using {}'.format(letter_dir, tempfile.gettempdir()))
+        letter_dir = Path(tempfile.gettempdir())
+    letter_dir = letter_dir / ('{:%Y%m%d-%H%M}'.format(datetime.now()))
+    try:
+        letter_dir.mkdir(exist_ok=True)
+    except:
+        print('Error creating {}'.format(letter_dir.absolute()))
+        return
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: no `return` needed here

File: students/shodges/lesson04/mailroom.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    for donor in donors.keys():
+        total = 0
+        for amount in donors[donor]: total += amount
+        letter_values = {'name': donor, 'amount': donors[donor][-1], 'total': total}
+        letter = letter_dir / (donor.replace(' ', '_') + '.txt')
+        with letter.open("w") as fileio: fileio.write(letter_template.format(**letter_values))
+        fileio.close()
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: you don't need to call `close()` with `with` statement

also, two separate lines

File: students/shodges/lesson04/mailroom.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    print('Saved letters in {}'.format(letter_dir.absolute()))
+
+def exit_mailroom():
+    return 1
```
Comment (natasha-aleksandrova) [inline students/shodges/lesson04/mailroom.py]: use `quit()` or `exit()` here

**Comment by theseanhodges** [comment]
Thanks @natasha-aleksandrova -- new PR is [359](https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/359)

**Pull request merged**